services/supervisor/src/transactions.rs

```diff
@@ -531,11 +668,115 @@ impl Supervisor {
         // Mark deploy as failed.
         schema
             .deploy_states
-            .put(&deploy_request, DeployState::Failed { height, error });
+            .put(&deploy_request, AsyncEventState::Failed { height, error });
 
         // Remove artifact from pending deployments: since we require
         // a confirmation from every node, failure for one node means failure
         // for the whole network.
         schema.pending_deployments.remove(&deploy_request.artifact);
     }
+
+    /// Confirms a local migration success by the given author's public key and checks
+    /// if all the confirmations are collected. If so, commits the migration.
+    /// If migration state hash differs from the expected one, migration fails though,
+    /// and `fail_migration` method is invoked.
+    fn confirm_migration(
+        &self,
+        mut context: CallContext<'_>,
+        request: MigrationRequest,
+        state_hash: Hash,
+        author: PublicKey,
+    ) -> Result<(), ExecutionError> {
+        let core_schema = context.data().for_core();
+
+        let mut schema = SchemaImpl::new(context.service_data());
+
+        let mut state = schema
+            .migration_states
+            .get(&request)
+            .expect("BUG: Attempt to confirm a migration which does not have a stored state");
+
+        // Verify that state hash does match expected one.
+        match state.add_state_hash(state_hash) {
+            Ok(()) => {
+                // Hash is OK, process further.
+            }
+            Err(error) => {
+                // Hashes do not match, rollback the migration.
+                drop(schema); // Required for the context reborrow.
+                let initiate_rollback = true;
+                return self.fail_migration(context, request, error, initiate_rollback);
+            }
+        }
+
+        // Check if we have enough confirmations to confirm the migration.
+        let confirmations = schema.migration_confirmations.confirm(&request, author);
+        let validator_count = core_schema.consensus_config().validator_keys.len();
```

Comment:
Can `consensus_config` change during the migration? If so, is it possible to rewrite this check to be sound?

Reply:
We can actualize the confirmations (as a set intersection of stored votes and current validator keys) every time we receive a confirmation, right before this check. Will this be enough?

Reply:
I think yes. The other option is to somehow restrict validator changes when there are pending migrations, but this is not quite good from the UX standpoint.